source/simulator/solver.cc

```diff
@@ -687,14 +713,32 @@ namespace aspect
         // processors
         catch (const std::exception &exc)
           {
+            signals.post_stokes_solver(*this, false, solver_history);
+
             if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
-              AssertThrow (false,
-                           ExcMessage (std::string("The iterative Stokes solver "
-                                                   "did not converge. It reported the following error:\n\n")
-                                       +
-                                       exc.what()))
-              else
-                throw QuietException();
+              {
+                // output solver history
+                std::ofstream f((parameters.output_directory+"solver_history.txt").c_str());
+                for (unsigned int i=0; i<solver_history.size(); ++i)
+                  {
+                    if (solver_history[i]<0)
+                      f << "\n";
+                    else
+                      f << i << " " << solver_history[i] << "\n";
+                  }
+                f.close();
+
+                std::cout << "See " << parameters.output_directory+"solver_history.txt"
+                          << " for convergence history." << std::endl;
+
+                AssertThrow (false,
+                             ExcMessage (std::string("The iterative Stokes solver "
+                                                     "did not converge. It reported the following error:\n\n")
```

Comment:
Put the error text above into the error message of the assertion -- this way everything comes out in the same place.